runtime/src/main/java/org/corfudb/runtime/clients/NettyClientProtobufRouter.java

```diff
@@ -0,0 +1,705 @@
+        final CompletableFuture<T> cf = new CompletableFuture<>();
+        outstandingRequests.put(thisRequest, cf);
+
+        // Write this message out on the channel
+        ByteBuf outBuf = PooledByteBufAllocator.DEFAULT.buffer();
+        ByteBufOutputStream requestOutputStream = new ByteBufOutputStream(outBuf);
+
+        try {
+            // Mark this message as Protobuf message (temporarily)
+            requestOutputStream.writeByte(API.PROTO_CORFU_MSG_MARK);
+            request.writeTo(requestOutputStream);
+            channel.writeAndFlush(outBuf, channel.voidPromise());
+            log.trace("Sent request message: {}", request.getHeader());
+        } catch (IOException e) {
+            log.warn("sendRequestAndGetCompletable[{}]: Exception occurred when sending request {}, caused by {}",
+                    request.getHeader().getRequestId(), request.getHeader(), e.getCause(), e);
+        } finally {
+            IOUtils.closeQuietly(requestOutputStream);
+        }
+
+        // Generate a benchmarked future to measure the underlying request
+        final CompletableFuture<T> cfBenchmarked = cf.thenApply(x -> {
+            MetricsUtils.stopConditionalContext(roundTripMsgContext);
+            return x;
+        });
+
+        // Generate a timeout future, which will complete exceptionally
+        // if the main future is not completed.
+        final CompletableFuture<T> cfTimeout =
+                CFUtils.within(cfBenchmarked, Duration.ofMillis(timeoutResponse));
+        Request finalRequest = request;
+        cfTimeout.exceptionally(e -> {
+            // CFUtils.within() can wrap different kinds of exceptions in
+            // CompletionException, just dealing with TimeoutException here since
+            // the router is not aware of it and this::completeExceptionally()
+            // takes care of others. This avoids handling same exception twice.
+            if (e.getCause() instanceof TimeoutException) {
+                outstandingRequests.remove(thisRequest);
+                log.debug("sendRequestAndGetCompletable: Remove request {} to {} due to timeout! Request:{}",
+                        thisRequest, node, finalRequest.getHeader());
+            }
+            return null;
+        });
+
+        return cfTimeout;
+    }
+
+    /**
+     * Send a one way message, without adding a completable future.
+     *
+     * @param reqBuilder The message builder for the request to send.
+     */
+    @Override
+    public void sendRequest(Request.Builder reqBuilder) {
+        // Get the next request ID
+        final long thisRequest = requestID.getAndIncrement();
+        // Set the base fields for this message.
+        Header header = reqBuilder.getHeaderBuilder()
+                .setClientId(API.getProtoUUID(parameters.getClientId()))
+                .setRequestId(thisRequest)
+                .build();
+        Request request = reqBuilder.setHeader(header).build();
```

Comment:
Here I pass the Request.Builder to avoid duplicate creation of the whole request message, but it still recreates the Header by invoking `getHeaderBuilder()`. To totally clear this kind of duplicate recreation, the only method I could think of now is pass two builders `sendRequest(Request.Builder reqBuilder, Header.Builder hdBuilder)`, which will make it too redundant and require lots of changes to upstream interfaces. 
Looking forward to hearing some suggestions here :)